src/v/raft/consensus.cc

```diff
@@ -1242,6 +1242,55 @@ consensus::remove_member(vnode node, model::revision_id new_revision) {
       });
 }
 
+std::optional<model::offset> consensus::adjusted_learner_initial_offset(
+  std::optional<model::offset> learner_start_offset) {
+    if (!learner_start_offset) {
+        return learner_start_offset;
+    }
+    /**
+     * We need to adjust the offset to the full batch boundary. This
+     * is required as the last offset included in the snapshot MUST be
+     * the last offset in a batch. If that wouldn't be the case Raft
+     * leader would not be able to recover follower as Raft delivers
+     * whole batches not individual records. We are looking for the last offset
+     * of a batch that is smaller than the requested learner_start_offset.
+     *
+     *
+     * The following invariant MUST hold:
+     * adjusted_offset <= last_included_offset
+     */
+
+    auto maybe_adjusted_offset = _log->index_lower_bound(*learner_start_offset);
+    while (maybe_adjusted_offset.has_value()
+           && maybe_adjusted_offset > learner_start_offset) {
+        maybe_adjusted_offset = _log->index_lower_bound(
+          model::prev_offset(*maybe_adjusted_offset));
+    }
```

Comment:
nit: feel free to ignore

One simpler solution probably is to make a log reader with start_offset as learner_start_offset and use the begin offset of the first batch returned by the reader (hides all the complexity, I think).

Reply:
you are 100% right, i was considering this but i was feeling the solution with a reader is much heavier

Reply:
Assuming the comment on `index_lower_bound` is correct, this loop may never finish. Scenario follows.

Batches [0..9]. [10..19], [20..29], all indexed.
`learner_start_offset` is {14}.
Before loop: `maybe_adjusted_offset` = {19}, as 10 <= 14 but 20 > 14.
Loop condition check: 19 > 14, pass.
Inside loop: `model::prev_offset(*maybe_adjusted_offset)` = 18;
`maybe_adjusted_offset` = {19}, as 10 <= 18 but 20 > 18.
Here we're stuck with `maybe_adjusted_offset`={19}.

Reply:
I changed this part to use reader, it is not often code path and reader version is much simpler and reliable